runtime/lua/vim/lsp.lua

```diff
@@ -260,10 +360,19 @@ end
 --- `ftplugin/<filetype_name>.lua` (See |ftplugin-name|)
 ---
 --- @param config vim.lsp.ClientConfig Configuration for the server.
---- @param opts vim.lsp.start.Opts? Optional keyword arguments
+--- @param opts vim.lsp.start.Opts? Optional keyword arguments.
 --- @return integer? client_id
 function lsp.start(config, opts)
   opts = opts or {}
+
+  if opts.attach == false then
+    local client_id, err = create_and_initialize_client(config)
+    if err and not opts.silent then
+      vim.notify(err, vim.log.levels.WARN)
+    end
+    return client_id
+  end
```

Comment:
Thoughts on doing something like this instead:

```lua
function lsp.start(config, opts)
  opts = opts or {}
  local reuse_client = opts.reuse_client or reuse_client_default
  local bufnr = resolve_bufnr(opts.bufnr)

  for _, client in pairs(all_clients) do
    if reuse_client(client, config) then
      if opts.attach == false then
        return client.id
      end
      if lsp.buf_attach_client(bufnr, client.id) then
        return client.id
      else
        return nil
      end
    end
  end

  local client_id, err = create_and_initialize_client(config)
  if err then
    if not opts.silent then
      vim.notify(err, vim.log.levels.WARN)
    end
    return nil
  end

  if opts.attach == false then
    return client_id
  end
  if client_id and lsp.buf_attach_client(bufnr, client_id) then
    return client_id
  end

  return nil
end
```

Basically remove the by-pass of `reuse_client` if `attach == false` and instead simply skip the `buf_attach_client` calls.

If one wanted to force creation of a new client they'd have to use `lsp.start(config, { reuse_client = function() return false end, attach = false })`, but for the mentioned use-case of restarting a client without attaching it to a possibly wrong buffer the existing `reuse_client_default` would do the right thing - start if it's missing, re-use it if not. So it would still be `vim.lsp.start(config, { attach = false })` for that use-case, but imho with more predictable behavior as it's "`lsp.start` except without attaching" instead of `lsp.start` except without attaching or re-using clients".